apps/src/music/utils/UniqueSounds.js

```diff
@@ -0,0 +1,75 @@
+export default class UniqueSounds {
+  constructor() {
+    this.previousUniqueSounds = null;
+  }
+
+  // Given some song data events, this function returns an array of the
+  // sound IDs in the order that we want to render them in the timeline UI.
+  // If available, it uses its own previous return value to ensure that
+  // sounds remain in the same row as last time.
+  //
+  // It does remove empty rows, so if a sound is removed, then everything
+  // can shift up.
+  //
+  // If a sound is replaced in subsequent calls, which will happen when a
+  // block has its sound changed in the dropdown, then the stickiness of
+  // all other sounds means that the new sound will replace the old one in
+  // the same row.
+
+  getUniqueSounds(songDataEvents) {
+    // Each unique sound gets its own row (and therefore color).
+    // If a sound was showing previously, then we'll attempt to keep
+    // it in the same row, though we won't have empty rows.
+
+    // First, generate a list of all unique sounds.
+    const uniqueSounds = [];
+    for (const songEvent of songDataEvents) {
+      const id = songEvent.id;
+      if (uniqueSounds.indexOf(id) === -1) {
+        uniqueSounds.push(id);
+      }
+    }
+
+    // This is the actual output.  It will be the same length as uniqueSounds.
+    // Fill it with undefined entries.
+    let outputSounds = new Array(uniqueSounds.length).fill(undefined);
+
+    // If we have a previous output from this function, then we'll attempt
+    // to keep those entries in the same rows as they were before.
+    if (this.previousUniqueSounds) {
+      // For each of those previous entries...
+      for (let i = 0; i < this.previousUniqueSounds.length; i++) {
```

Comment:
Would it be better to iterate through `uniqueSounds` here instead and check `this.previousUniqueSounds.indexOf(uniqueSounds[i])`? That way you're not checking sounds that may no longer be in use?

Reply:
Great question.  I might be missing a good shortcut, but given we do two passes -- one to pull over sounds into their existing slots, and another to fill in the gaps with the remaining sounds -- I think the resulting code for that first pass is not greatly different, and is in fact is a tiny bit longer:

```
      // For each of the current sounds...
      for (let i = 0; i < currentUniqueSounds.length; i++) {
        // ...see if it already has a slot...
        const previousSlot = this.previousUniqueSounds.indexOf(
          currentUniqueSounds[i]
        );
        // ...and if so, put it back in there.
        if (previousSlot !== -1) {
          outputSounds[previousSlot] = this.previousUniqueSounds[previousSlot];
        }
      }
```

Reply:
Yeah, that's fair. I suppose another slight optimization you could make here is to push any new sounds that weren't the previous list into a second array, and then iterate through that after, instead of going through the whole list again? At this scale though it's not gonna make a huge difference, so whatever is easiest.

Reply:
Oh that's a great idea, and is now implemented.